File: scripts/generate-markdown.py
import requests
import sys
import os

current_path = os.path.dirname(os.path.abspath(__file__))

URL_Newsletter = "https://dnw-newsletter.azurewebsites.net/api/v1/"
URL_API = "https://dnw-api.azurewebsites.net/api/v1/"

newsletters = requests.get('%snewsletters' % URL_Newsletter).json()
linksData = {}

fpost = open("%s/post.md" % current_path, "r")
post_tmpl = fpost.read()

if not(newsletters["success"]):
    print("No Newsletters Found")
    sys.exit()


def makeFolders(year, week):
    yearPath = "%s/../src/newsletters/%s" % (current_path, year)
    if not os.path.exists(yearPath):
        os.makedirs(yearPath)
    weekPath = "%s/%s" % (yearPath, week)
    if not os.path.exists(weekPath):
        os.makedirs(weekPath)
    return weekPath


for newsletter in newsletters["data"]:
    year = newsletter["year"]
    week = newsletter["week"]

    weekPath = makeFolders(year, week)
    linksJson = {}
    if not year in linksData:
        linksData[year] = {}
    if not week in linksData[year]:
        linksData[week] = {}
        linksJson = requests.get(
            '%slinks?category=&week=%s&year=%s' % (URL_API, week, year)).json()

    linksJson = linksJson["data"]["links"]
    for linkJson in linksJson:

        postPath = "%s/%s.md" % (weekPath, linkJson["slug"])
        if not os.path.exists(postPath):
            newPostData = requests.get('%slinks/%s' %
                                       (URL_API, linkJson["slug"])).json()
            if newPostData["success"]:
                new_post = post_tmpl % (
                    linkJson["_id"],
                    linkJson["title"],
                    linkJson["url"],
                    linkJson["category"],
                    linkJson["slug"],
                    linkJson["user"]["_id"],
                    linkJson["createdOn"],
                    str.join(",", (("'%s'" % tag)
                                   for tag in linkJson["tags"])),
                    newPostData["data"]["content"]
                )
                f = open(postPath, "x")
                f.write(new_post)

File: scripts/generate-markdown.js
const fs = require('fs')
const request = require('request')
const path = require('path')
const config = require('../conf')
const weeklyCalendarHelper = require('weekly-calendar-helper')

const postTmpl = fs.readFileSync(path.join(__dirname, 'post.md')).toString()

const URL_API = config.apiDomain

async function getLinks(week, year) {
  return await new Promise((resolve, reject) => {
    request(
      `${URL_API}links?category=&week=${week}&year=${year}`,
      function (error, response, body) {
        const linkJson = JSON.parse(body)
        linkJson.data.links
        resolve(linkJson.data.links)
      }
    )
  })
}

async function createPost(week, year, link) {
  const weekPath = path.join(__dirname, `../src/newsletters/${year}/${week}`)
  const postPath = `${weekPath}/${link['slug']}.md`
  if (fs.existsSync(postPath)) {
    return
  }

  return new Promise((resolve, reject) => {
    request(
      `${URL_API}links/${link['slug']}`,
      function (error, response, body) {
        try {
          const linkJson = JSON.parse(body)
          const postData = postTmpl
            .replace('_id: %s', `_id: "${link['_id']}"`)
            .replace(
              'title: "%s"',
              `title: "${link['title'].replace(/"/g, '')}"`
            )
            .replace("url: '%s'", `url: '${link['url']}'`)
            .replace('category: %s', `category: ${link['category']}`)
            .replace("slug: '%s'", `slug: '${link['slug']}'`)
            .replace('user_id: %s', `user_id: ${link['user']['_id']}`)
            .replace(
              "username: '%s'",
              `username: '${link['user']['username']}'`
            )
            .replace("createdOn: '%s'", `createdOn: '${link['createdOn']}'`)
            .replace('tags: [%s]', `tags: [${link['tags'].join(',')}]`)
            .replace('**CONTENT**', `${linkJson.data['content']}`)

          fs.writeFileSync(postPath, postData)
          resolve()
        } catch (ex) {
          reject()
        }
      }
    )
  })
}

async function generate() {
  let currentWeek = 20
  let currentYear = 2012

  const now = weeklyCalendarHelper.baseHelper.getUtcNow()
  const nowWeek = weeklyCalendarHelper.weekHelper.getWeekNumber(now)
  const lastWeek = weeklyCalendarHelper.weekHelper.getWeekNumber(now)
  const lastYear = now.getFullYear()

  let count = 0

  let lastOne = false
  let breakIt = false

  while (!breakIt) {
    if (
      !fs.existsSync(path.join(__dirname, `../src/newsletters/${currentYear}`))
    ) {
      fs.mkdirSync(path.join(__dirname, `../src/newsletters/${currentYear}`))
    }
    let folderExists = fs.existsSync(
      path.join(__dirname, `../src/newsletters/${currentYear}/${currentWeek}`)
    )

    if (!folderExists) {
      fs.mkdirSync(
        path.join(__dirname, `../src/newsletters/${currentYear}/${currentWeek}`)
      )
    }
    if (
      !folderExists ||
      (now.getFullYear() === currentYear && nowWeek - 1 === currentWeek) ||
      (now.getFullYear() === currentYear && nowWeek === currentWeek)
    ) {
      const links = await getLinks(currentWeek, currentYear)
      for (key in links) {
        await createPost(currentWeek, currentYear, links[key])
      }
    }
    const yearWeeks = weeklyCalendarHelper.weekHelper.weeksInYear(currentYear)
    currentWeek++
    if (yearWeeks < currentWeek) {
      currentWeek = 1
      currentYear++
    }
    if (lastOne) {
      breakIt = true
    }
    if (lastYear === currentYear && lastWeek === currentWeek) {
      lastOne = true
    }

    count++
    if (count > 2500) {
      break
    }
  }
  return
}

exports.generateMarkdown = async function () {
  await generate()
}

const args = process.argv
if (args.length > 0 && args[2] == 'run') {
  generate()
}
